domain/src/com/thoughtworks/go/domain/materials/perforce/P4MaterialUpdater.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2017 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.domain.materials.perforce;
+
+import com.thoughtworks.go.config.materials.perforce.P4Material;
+import com.thoughtworks.go.domain.BuildCommand;
+import com.thoughtworks.go.domain.materials.RevisionContext;
+import com.thoughtworks.go.util.GoConstants;
+import com.thoughtworks.go.util.StringUtil;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.thoughtworks.go.domain.BuildCommand.*;
+import static java.lang.Long.parseLong;
+import static java.lang.String.format;
+
+public class P4MaterialUpdater {
+    private P4Material material;
+    private String clientName;
+
+    public P4MaterialUpdater(P4Material material) {
+        this.material = material;
+    }
+
+    public BuildCommand updateTo(String baseDir, RevisionContext revisionContext) {
+        String workingDir = material.workingdir(new File(baseDir)).getPath();
+        String revision = revisionContext.getLatestRevision().getRevision();
+        this.clientName = material.clientName(new File(workingDir));
+
+        return compose(
+                secret(material.getPassword()),
+                loginIfUsingTickets(),
+                constructClient(workingDir, clientName),
+                cleanWorkingDir(workingDir),
+                echo(format("[%s] Start updating %s at revision %s from %s", GoConstants.PRODUCT_NAME, material.updatingTarget(), revision, material.getServerAndPort())),
+                sync(workingDir, revision),
+                echo(format("[%s] Done.\n", GoConstants.PRODUCT_NAME))
+
+        );
+    }
+
+    private BuildCommand loginIfUsingTickets() {
+       if (material.getUseTickets() && !StringUtil.isBlank(material.getPassword())) {
+           return exec("p4", "login").setExecInput(material.getPassword()).setCommandEnvVars(envVars());
+       } else {
+           return noop();
+       }
+    }
+
+    private Map<String, String> envVars() {
+        Map<String, String> env = new HashMap<>();
+        env.put("P4PORT", material.getServerAndPort());
+        env.put("P4CLIENT", clientName);
+        if (!StringUtil.isBlank(material.getUserName())) {
+            env.put("P4USER", material.getUserName());
+        }
+        if (material.getUseTickets() && !StringUtil.isBlank(material.getPassword())) {
```

Comment:
@ibnc I've been trying to verify and merge all of the PRs. I've been able to merge (locally) and verify that SVN, Mercurial and Package Material are all fine. This one is giving me some trouble, though. Apart from the issue I mentioned about the `P4PASSWD`, I'm struggling to get this to actually work. Here is my setup:

```
<pipeline name="Test">
  <materials>
    <p4 port="workshop.perforce.com:1666" username="abc_1234" password="password" dest="mat1" materialName="p4a">
      <view><![CDATA[//guest/abc_1234/testp41/main... //client1/...]]></view>
    </p4>
  </materials>
  <stage name="defaultStage">
    <jobs>
      <job name="defaultJob" runOnAllAgents="true">
        <tasks>
          <exec command="ls">
            <arg>-laR</arg>
            <runif status="passed" />
          </exec>
        </tasks>
      </job>
    </jobs>
  </stage>
</pipeline>
```

With the old code, the checkout actually happens and the files are visible. With this new code, the files aren't even checked out. Build passes, though. I wrote an ugly little wrapper over P4 to log everything that's happening. It looks like this:

```sh
#!/bin/bash

file="/tmp/p4.commands.${PPID}"
out="$(mktemp)"
err="$(mktemp)"

trap 'rm -f "$out" "$err"' EXIT

echo "---------------------" >>"$file"
echo "Command: p4 $@" >>"$file"
echo "PWD: $(pwd)" >>"$file"
env | grep '^P4' | sed -e 's/^/  /' >>"$file"
echo >>"$file"

if test -s /dev/stdin
then
   input="$(cat)"
   if [ -n "$input" ]; then
       echo -e "INPUT:\n$(echo "$input" | sed -e 's/^/    | /')" >>"$file"
   fi
fi

echo "$input" | exec /usr/local/bin/p4.old "$@" >"$out" 2>"$err"

echo -e "OUTPUT:\n$(cat "$out")\n" >>"$file"
echo -e "ERR:\n$(cat "$err")" >>"$file"

cat "$out"
cat "$err" >&2
```

I've attached the output of this for the old agent and the new one (I mean, p4 over buildsession -
 P4MaterialUpdater).

If you have time to take a look, that's fine. If not, I can take a look and fix it. Let me know. Thanks!

[p4.new.txt](https://github.com/gocd/gocd/files/1134052/p4.new.txt)
[p4.old.txt](https://github.com/gocd/gocd/files/1134051/p4.old.txt)

Not sure if it's a setup issue on my side, but the fact that this is a "run on all agents" job and it runs properly on the old agent is making me think it's not a setup issue.